src/test/java/net/jonathangiles/tools/teenyhttpd/StoreController.java

```diff
@@ -0,0 +1,61 @@
+package net.jonathangiles.tools.teenyhttpd;
+
+import net.jonathangiles.tools.teenyhttpd.annotations.*;
+import net.jonathangiles.tools.teenyhttpd.implementation.ResponseEntity;
+
+import java.util.List;
+import java.util.ArrayList;
+
+@Path("/store")
+public class StoreController {
+
+    private final List<Product> productList;
+
+    public StoreController() {
+        productList = new ArrayList<>();
+        productList.add(new Product(1, "Apple", 100));
+        productList.add(new Product(2, "Banana", 200));
+    }
+
+    @Get
+    public String sayHello() {
+        return "Hello, World!";
+    }
+
+    @Get("/products")
+    public List<Product> get() {
+        return productList;
+    }
+
+    @Get("/product/:id")
+    public ResponseEntity<Product> getProduct(@PathVariable("id") int id) {
+
+        for (Product product : productList) {
+            if (product.getId() == id) {
+                return ResponseEntity.ok(product);
+            }
+        }
+
+        return ResponseEntity.notFound();
+    }
```

Comment:
There are two things here:

1. Because `ResponseEntity` is now implementation, users should not use it. We hide things from people that are in the implementation package.
2. I misunderstood something previously, because the existing `Response` type is really from the servers perspective - it is its way of responding to the client by writing values *into* the response. Conversely, your `ResponseEntity` is the response from the clients perspective, so it is more focused on reading the values *out* of the response.

This is slightly annoying! We have two things that are very similar, but not quite the same. We need different types, one focused on writing, the other focused on reading - but they need unique names. I'm not sure ResponseEntity is the right name, and I now understand better that `ResponseEntity` (or whatever it is called) should not implement the `Response` interface.

Reply:
Actually, I'm just completely wrong. I shouldn't respond to GitHub issues before my morning coffee :-)

The real crux of the issue is that my Response type is focused on writing some stream of bytes into a socket, whereas your ResponseEntity type is focused on working with higher-level concepts (in particular, supporting generic types). We need to reconcile these two use cases.

Reply:
I wonder if we do the following:

1. Introduce a `TypedResponse<T>` interface that implements `Response`, with `create(T)` method and a `T getValue()` method.
2. Have a `TypedResponseImpl<T>` class in the implementation package, which is the concrete type that the `TypedResponse.create(T)` method uses.
3. Get rid of `ResponseEntity` and `MessageConverter` APIs, and just move it into the `TypedResponse` API?

Reply:
## TypedResponse

`TypedResponse<T>` represents the entire HttpResponse as it holds Headers, Body and StatusCode, this is why is part of the public API, an interface is actually not required 'cause `TypedResponse` is basically a wrapper for (Header, Body & StatusCode) So this should not implement the Response interface ('cause Response is in a lower level). 
How it's handled depends on `EndpointHandler` that will use the most suitable `MessageConverter` to handle its body `T`.


For Example:

> ContentType: application/json
```
TypedResponse<Pet> -> [MessageConverter : JsonConverter] -> StringResponse
```

> ContentType: text/plain
```
TypedResponse<String> -> [MessageConverter : DefaultConverter] -> StringResponse
```

> ContentType: application/x-protobuf
```
TypedResponse<String> -> [MessageConverter : ProtobufConverter] -> ByteResponse
```

## MessageConverter

The `MessageConverter` interface allows the user to specify how an entity should be handled.

for example:

```java
class GsonMessageConverter implements MessageConverter {

        final Gson gson = new Gson();

        @Override
        public String getContentType() {
            return "application/json";
        }

        @Override
        public void write(Object value, BufferedOutputStream dataOut) throws IOException {
            dataOut.write(gson.toJson(value).getBytes());
        }
    }
```

```java
TeenyApplication.start()
                .registerMessageConverter(new GsonMessageConverter())
                .register(new StoreController());
```

> Now everything related to the content-type `application/json` will be handled by  GsonMessageConverter, we can also add a method `boolean isSupported(java.lang.reflect.Type type)` to filter which types are supported by this converter.


```java
@Override
        public void write(Object value, BufferedOutputStream dataOut) throws IOException {
            if( object instanceOf ProtoPet){
               ((ProtoPet)value).writeTo(dataOut);
            }
            throws new IllegalStateException("Oh no... unable to write protobuf of type: "+value.getClass());
        }
```

> Teeny protobuf example

---

meanwhile in my head

![Blank diagram-2](https://github.com/JonathanGiles/TeenyHttpd/assets/25180512/722689f3-81c3-48e1-b215-bb0be395dfb6)


Maybe it's too much effort :S

Reply:
As I've experienced on TeenyHttpd, things churn a lot over time. I'm sure the same will happen here too! Let's get your stuff as polished as possible, and then we can get it merged and iterate over time.